### PR TITLE
Update JSX intrinsic element test to match babel’s

### LINKED ISSUE #19945: JSXElement name that starts with an underscore is assumed to be an intrinsic element
Typescript and babel both have a very similar test for deciding whether or not a JSXElement is a intrinsic element or a component.

- babel: [packages/babel-types/src/react.js#L5-L7](https://github.com/babel/babel/blob/25bde88/packages/babel-types/src/react.js#L5-L7)
  ```js
  /^[a-z]|-/.test(tagName)
  ```
- Typescript: [src/compiler/utilities.ts#L2436](https://github.com/Microsoft/TypeScript/blob/b8fbf88/src/compiler/utilities.ts#L2436)
  ```js
  tagName[0] === tagName[0].toLowerCase()
  ```

Typescript’s test has one fatal flaw, however: it relies on a `toLowerCase` comparison but does not verify that the first character is a letter. I encountered an issue where a JSXElement starting with an underscore was incorrectly assumed to be an intrinsic element.

My proposal is that the Typescript JSXElement name test be more inline with the babel JSXElement name test. I’ll whip up a PR as soon as I figure out how things are done around here.

**TypeScript Version:**  2.6.1

**Code**
```ts
import { Box as _Box } from 'jsxstyle';
<_Box />;
```

**Expected behavior:**
`_Box` is treated like a component instead of an intrinsic element.

**Actual behavior:**
`error TS2339: Property '_Box' does not exist on type 'JSX.IntrinsicElements'.`

For reference: https://github.com/smyte/jsxstyle/issues/82#issuecomment-341506355

### PR DESCRIPTION
**Pull request by meyer**
This PR changes the intrinsic element check in [src/compiler/utilities.ts#L2436](https://github.com/Microsoft/TypeScript/blob/b8fbf88/src/compiler/utilities.ts#L2436) to use babel’s instead.

Fixes #19945

Babel’s JSXElement name test: [babel-types/src/react.js#L5-L7](https://github.com/babel/babel/blob/25bde88/packages/babel-types/src/react.js#L5-L7)

This PR undoes some of the changes added by #16915 and makes the `doubleUnderscoreReactNamespace` test redundant. Might be good to get input from @weswigham (?)

### CODE REVIEW

**Comment by meyer** [comment]
@RyanCavanaugh updated, thanks!

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -2434,10 +2434,7 @@ namespace ts {
     }
 
     export function isIntrinsicJsxName(name: __String | string) {
-        // An escaped identifier had a leading underscore prior to being escaped, which would return true
-        // The escape adds an extra underscore which does not change the result
-        const ch = (name as string).substr(0, 1);
-        return ch.toLowerCase() === ch;
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: `ch.toLowerCase() === ch || ch === "_"` ?

Comment (mhegazy) [reply src/compiler/utilities.ts]: aah.. that is the opposite.. sorry..

Comment (mhegazy) [reply src/compiler/utilities.ts]:
```ts
const ch = (name as string).charCodeAt(0);
return (ch >= CharacterCodes.a && ch <= CharacterCodes.z) ||
 ch == CharacterCodes._ ||
 (name as string).indexof("-") > 0;
```

Comment (mhegazy) [reply src/compiler/utilities.ts]: and avoid the creation of a regexp on every invocation.

Comment (meyer) [reply src/compiler/utilities.ts]: @mhegazy good catch, thanks!

Comment (meyer) [reply src/compiler/utilities.ts]: oops, missed the comment before. Updating…

**Pull request merged**